### ISSUE TITLE
Incorrect SSZ encoding for SignedBuilderBid

### ISSUE THREAD

**Issue by jtraglia** [attestantio/go-builder-client#9]
Hello, I noticed a difference between mev-boost & this package when SSZ encoding a `SignedBuilderBid` object. In the output below, `Bid1` is the SSZ that mev-boost generates & `Bid2` is the SSZ that this package generates. Near the beginning, there's a missing chunk of 32 bytes (spoiler, this is the value field).

```
signing_test.go:160: Bid1: 0x540000000000000000000000000000000000000000000000000000000000000000000000b5246e299aeb782fbc7c91b41b3284245b1ed5206134b0028b81dfb974e5900616c67847c2354479934fc4bb75519ee10544e2170998060d9560fdbf8f263a08c0a209211569a0560138522b84805abc0000000000000000000000000000000000000000cded53d652660a91bfe6f5dfb017204a4cdd1598a07116b2cdea1586d603d01c56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b42100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000d60955dc7f0cc7bf28d7e6c6f4859081f3a6df5ef4f70e05d70d8282bac20c6c4fa70e000000000080c3c901000000000000000000000000d051ed6200000000180200000700000000000000000000000000000000000000000000000000000000000000ea33078b00e6b2926f45ed6d3190a3a6ada75cee342f600cf22fa02a9a2edcb77ffe241ea60187fdb0187bfa22de35d1f9bed7ab061d9401fd47e34a54fbede1466c617368626f747320666c617368626c6f636b
signing_test.go:166: Bid2: 0x54000000b5246e299aeb782fbc7c91b41b3284245b1ed5206134b0028b81dfb974e5900616c67847c2354479934fc4bb75519ee10544e2170998060d9560fdbf8f263a08c0a209211569a0560138522b84805abc0000000000000000000000000000000000000000cded53d652660a91bfe6f5dfb017204a4cdd1598a07116b2cdea1586d603d01c56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b42100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000d60955dc7f0cc7bf28d7e6c6f4859081f3a6df5ef4f70e05d70d8282bac20c6c4fa70e000000000080c3c901000000000000000000000000d051ed6200000000180200000700000000000000000000000000000000000000000000000000000000000000ea33078b00e6b2926f45ed6d3190a3a6ada75cee342f600cf22fa02a9a2edcb77ffe241ea60187fdb0187bfa22de35d1f9bed7ab061d9401fd47e34a54fbede1466c617368626f747320666c617368626c6f636b
signing_test.go:167: Value: []
```

<details>
<summary>Code used to generate this</summary>

```go
bidStr := `{"message":{"header":{"parent_hash":"0x0544e2170998060d9560fdbf8f263a08c0a209211569a0560138522b84805abc","fee_recipient":"0x0000000000000000000000000000000000000000","state_root":"0xcded53d652660a91bfe6f5dfb017204a4cdd1598a07116b2cdea1586d603d01c","receipts_root":"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421","logs_bloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","prev_randao":"0xd60955dc7f0cc7bf28d7e6c6f4859081f3a6df5ef4f70e05d70d8282bac20c6c","block_number":"960335","gas_limit":"30000000","gas_used":"0","timestamp":"1659720144","extra_data":"0x466c617368626f747320666c617368626c6f636b","base_fee_per_gas":"7","block_hash":"0xea33078b00e6b2926f45ed6d3190a3a6ada75cee342f600cf22fa02a9a2edcb7","transactions_root":"0x7ffe241ea60187fdb0187bfa22de35d1f9bed7ab061d9401fd47e34a54fbede1"},"value":"0","pubkey":"0xb5246e299aeb782fbc7c91b41b3284245b1ed5206134b0028b81dfb974e5900616c67847c2354479934fc4bb75519ee1"},"signature":"0xa775df980d589a87b234cf36b94fbcd40540ab1dffb752a013c02f636d85db60023f7e9d883de8cfdbfd94e0e3b598c01429fee50a5cb8d9fce50557baec2e9f81268f14f4f044b44b1238b7945201f036036d1a25d60e681f3737d4ef3b54b6"}`

// Decode with mev-boost types.
bid1 := new(types.SignedBuilderBid)
require.NoError(t, utils.DecodeJSON(strings.NewReader(bidStr), bid1))
bid1Ssz, err := bid1.Message.MarshalSSZ()
require.NoError(t, err)
t.Log("Bid1:", hexutil.Encode(bid1Ssz))

// Decode with attestant types.
bid2 := new(bellatrix_api.SignedBuilderBid)
require.NoError(t, utils.DecodeJSON(strings.NewReader(bidStr), bid2))
bid2Ssz, err := bid2.Message.MarshalSSZ()
require.NoError(t, err)
t.Log("Bid2:", hexutil.Encode(bid2Ssz))
t.Log("Value:", bid2.Message.Value.Bytes())
```

</details>

The problem is with the `Value` field, which is a `uint256.Int` type:
https://github.com/attestantio/go-builder-client/blob/fe1a8d4e7677d47e00e8fe61a752548e45585498/api/bellatrix/builderbid.go#L34

In Holiman's [`uint256`](https://pkg.go.dev/github.com/holiman/uint256) package, the `Bytes` method unintuitively returns the minimal number bytes, instead of 32 bytes like one might expect. This confuses `fastssz` which uses this method:
https://github.com/attestantio/go-builder-client/blob/fe1a8d4e7677d47e00e8fe61a752548e45585498/api/bellatrix/builderbid_encoding.go#L28-L33

It should have generated something like this:

```diff
--- a/api/bellatrix/builderbid_encoding.go
+++ b/api/bellatrix/builderbid_encoding.go
@@ -26,11 +26,11 @@ func (b *BuilderBid) MarshalSSZTo(buf []byte) (dst []byte, err error) {
        offset += b.Header.SizeSSZ()
 
        // Field (1) 'Value'
-       value := b.Value.Bytes()
+       value := b.Value.Bytes32()
        for i, j := 0, len(value)-1; i < j; i, j = i+1, j-1 {
                value[i], value[j] = value[j], value[i]
        }
-       dst = append(dst, value...)
+       dst = append(dst, value[:]...)
 
        // Field (2) 'Pubkey'
        dst = append(dst, b.Pubkey[:]...)
```

Because this is generated code, I'm not entirely sure what the proper solution should be. I don't believe we can hint to `fastssz` about which method to use there. A possible option is to wrap the `uint256.Int` method and override the `Bytes` method. Or create your own basic `uint256` class, like mev-boost does.

**Comment by jtraglia** [comment]
Also worth mentioning, this applies to a few other places too:

https://github.com/attestantio/go-builder-client/blob/fe1a8d4e7677d47e00e8fe61a752548e45585498/api/bellatrix/builderbid_encoding.go#L117-L122

https://github.com/attestantio/go-builder-client/blob/fe1a8d4e7677d47e00e8fe61a752548e45585498/api/capella/builderbid_ssz.go#L29-L34

https://github.com/attestantio/go-builder-client/blob/fe1a8d4e7677d47e00e8fe61a752548e45585498/api/capella/builderbid_ssz.go#L118-L123

https://github.com/attestantio/go-builder-client/blob/fe1a8d4e7677d47e00e8fe61a752548e45585498/api/v1/bidtrace_encoding.go#L44-L49

https://github.com/attestantio/go-builder-client/blob/fe1a8d4e7677d47e00e8fe61a752548e45585498/api/v1/bidtrace_encoding.go#L136-L141

**Comment by mcdee** [comment]
If I'm reading this correctly, should `uint256` always be a 64-byte value, and if is it an issue to address in `fastssz`?  Or is there any situation where `uint256.Bytes()` would be a valid encoding?

**Comment by jtraglia** [comment]
According to [the SSZ spec](https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md#uintn), a `uint256` value should always encoded as a 32-byte value. And yeah, I think it should be fixed in `fastssz` if possible. I'm not certain, but I believe `uint256.Bytes()` is valid for something, but not SSZ.

**Comment by jtraglia** [comment]
> Or is there any situation where uint256.Bytes() would be a valid encoding?

Sorry, I misunderstood your question. Yes, it would be valid when the lowest (little-endian) byte is non-zero.

**Comment by mcdee** [comment]
Yeah sorry the question wasn't very precise; I was wondering if there was any situation where an output with fewer than 32 bytes would be valid SSZ, but I suspect not.

**Comment by metachris** [comment]
That's a great find, thanks @jtraglia! 🕵️ 

How do we proceed from here?

**Comment by jtraglia** [comment]
@mcdee I'm having trouble generating `BuilderBid` with sszgen. How was this done?

```
jtraglia at desktop in ~/Projects/attestantio/go-builder-client/api/bellatrix (master●) 
$ go generate    
...
[ERR]: failed to encode BuilderBid: could not find struct with name 'Int'
generate.go:18: running "sszgen": exit status 1
```

With `../../../../holiman/uint256` added to the list of includes, it has this error:

```
jtraglia at desktop in ~/Projects/attestantio/go-builder-client/api/bellatrix (master●) 
$ go generate
...
[ERR]: failed to encode BuilderBid: failed to encode Int: unexpected mismatch between ssz-size and array fixed size, name=Int, ssz-size=32, fixed=4
generate.go:18: running "sszgen": exit status 1
```

**Comment by mcdee** [comment]
Are the SSZ encoded versions in use today?

If so, if someone can supply both correct and incorrect encodings for a `SignedBuilderBid` I can take a look at fixing this in a two-step process as follows:

1. change decoding to use either format
2. change encoding to use correct format (i.e. fixed 32-byte size for uint256)

Otherwise, I can just change encoding and decoding to use the correct format.

**Comment by metachris** [comment]
well, both the relay and mev-boost and the builder uses go-builder-client@v0.2.8, so all of them use the incorrect encoding. for `SignedBuilderBid` and `BidTrace`.

I think the best way forward will be to simply fix everything that uses `uint256.Int`.

**Comment by mcdee** [comment]
> I think the best way forward will be to simply fix everything that uses `uint256.Int`.

I can do that, but that will create an incompatibility between versions.  If you are comfortable with builders and relays being able to upgrade in lock-step then I can go ahead with this as the fix.

**Comment by jtraglia** [comment]
Okay, so I have some really good news. The hash tree root is different ONLY when the message.value is zero. For non-zero values, even though the SSZ is incorrect, the HTR value is correct. And because the relays reject zero-value bids, this should never happen in practice. So it should be safe to fix & update mev-boost, mev-boost-relay, and builders. This explains why everything's working fine right now, because otherwise I'd expect there to be signature disagreements.

**Comment by mcdee** [comment]
Sounds like a simple fix should do the job, then.  Please could someone supply me with a few correctly-encoded objects so that I can add them to the tests?

**Comment by holiman** [comment]
@karalabe didn't you already solve this, back when you noticed the same problem in the 4844-implementation? 

cc @MariusVanderWijden

**Comment by holiman** [comment]
Yeah, so one part of the solution was to add the ssz interfaces to `uint256`: 
https://github.com/holiman/uint256/pull/126/files

Another part was to properly invoke the interface, IIRC some other change was made to the ssz-library to accommodate this.

**Comment by mcdee** [comment]
This issue should be addressed in #10, @jtraglia @metachris please could you take a look at this and see if you're happy with it?

The PR is quite large because of other changes in the `go-eth2-client` repo, but the main files to look at are:

- api/bellatrix/builderbid_encoding.go
- api/capella/builderbid_ssz.go
- api/v1/bidtrace_encoding.go

**Comment by holiman** [comment]
@mcdee might have been better to use the ssz marshalling methods, `MarshalSSZTo`, instead of using `b.Value.Bytes32()` and for-looping the endian-ness thingy

https://github.com/holiman/uint256/pull/126/files#diff-4caa05d97eba477e577a7721860f68293d7371001a42747af73e647b581a5e9bR470

https://github.com/attestantio/go-builder-client/pull/10/files#diff-0b7d53e341f9d6f752684cd80200ceafb2143527ce57365c62442f499aae8cf1R119

So instead of 
```golang
	value := b.Value.Bytes32()
	for i, j := 0, 31; i < j; i, j = i+1, j-1 {
		value[i], value[j] = value[j], value[i]
	}
	hh.PutBytes(value[:])
```
You could have done 

```golang
	value, err := b.Value.MarshalSSZ()
	// perhaps handle error
	hh.PutBytes(value[:])
```
